### ISSUE TITLE
Clang doesn't warn on some cases of `-Wimplicit-fallthrough` that GCC catches

### ISSUE THREAD

**Issue by ilovepi** [llvm/llvm-project#99901]
When building LLD with GCC @4vtomat found that we don't emit diagnostics for certain types of nested switch statements. This was addressed in source in #99762, but I think we should probably catch this in Clang. If this is an intentional difference due to interpretation of some standard, then we should probably document that somehow.

The following godbolt example illustrates this https://godbolt.org/z/eGETvo5jq

clang -c example.cpp -Werror -Wimplicit-fallthrough
```c++
// The following run lines will succeed in clang and fail w/ gcc
// RUN: clang -c example.cpp -Werror -Wimplicit-fallthrough
// RUN: gcc -c example.cpp -Werror -Wimplicit-fallthrough

enum color { Blue, Red };
bool foo(color a, color b) {
    bool different = false;
    switch (a) {
    case Blue:
        switch (b) {
        case Blue:
            return different;
        case Red:
            different = true;
            return different;
        };
    case Red:
        switch (b) {
        case Blue:
            return different;
        case Red:
            different = true;
            return different;
        };
    };
    return different;
}
```

**Comment by 4vtomat** [comment]
Seems there are still some places that have this problem.
